text/0089-tenant-token.md

```diff
@@ -0,0 +1,125 @@
+- Title: Tenant Tokens
+- Start Date: 2021-10-15
+- Specification PR: [#89](https://github.com/meilisearch/specifications/pull/89)
+- Discovery Issue: [#51](https://github.com/meilisearch/product/issues/51)
+
+# Tenant Token
+
+## 1. Functional Specification
+
+### I. Summary
+
+The SDKs can generate `Tenant tokens` inheriting from a MeiliSearch's `API key` generated to force filters during a search for an end-user. A `Tenant token` is generated on the user server-side code to be used by an end-user. It allows users to have multi-tenant indexes and thus restricts access to documents depending on the end-user making the search request.
+
+### II. Motivation
+
+`Tenant tokens` are introduced to solve multi-tenant use-cases. By providing the management of API Keys, we remove one of the last major deal-breakers that makes users not choose MeiliSearch as a solution despite all our advantages.
+
+Users regularly request Multi-Tenant indexes. Users today need to set up workarounds to meet this need. Some of them implement reverse-proxy or managed authentication systems like Hasura or Kong to filter what can and cannot be read. Others decide to use server code as a facade to implement the access restriction logic before requesting MeiliSearch. It is difficult to maintain, less efficient, and requires necessary skills that not everyone has.
+
+### III. Glossary
+
+| Term               | Definition |
+|--------------------|------------|
+| Master Key         | This is the master key that allows the creation of other API keys. The master key is defined by the user when launching MeiliSearch. |
+| API Key            | API keys are stored and managed from the endpoint `/keys` by the master key holder. These are the keys used by the technical teams to interact with MeiliSearch at the level of the frontend/backend code. |
+| Tenant Token     | These tokens are not stored and managed by a MeiliSearch instance. They are generated for each end-user by the backend code from a MeiliSearch API Key. They are used by the end-users to only search the documents belonging to them. |
```

Comment:
```suggestion
| Tenant Token     | These tokens are not stored and managed by a MeiliSearch instance. They are generated for each end-user by the backend code from a MeiliSearch API Key. 
They are used by end-users to search only for documents to which the tenant token gives them access. |
```
Did the documents really belong to the end-users or do they just have access to them? Not sure it's better but it still confusing for me.

Reply:
Yes, the documents belong to them or in any case should be accessible only to them. By adding the tenant tokens it allows us to overcome two pain-full workarounds to improve the quality of life of a MeiliSearch user. Here is what they have to do to be sure that a document is really accessible to one tenant.

1 - Make an index by tenant (user, company, ...), which can be difficult to manage.
2 - To have a single index by adding a proxy that will inject the forced filter from its backend which forces to have network round trips that will become useless with the tenant token once generated since it will be able to directly exchange with MeiliSearch.

Reply:
Thanks for the clarification 😊